scripts/generate_matrix.py

```diff
@@ -0,0 +1,273 @@
+        if version < min_version:
+            continue
+
+            # Get 'requires_python' for this version
+        requires_python = None
+        for file_info in release_files:
+            if "requires_python" in file_info:
+                requires_python = file_info["requires_python"]
+                if requires_python:
+                    break
+
+        if requires_python is None:
+            # Skip versions without 'requires_python'
+            continue
+
+        key = (version.major, version.minor)
+        # Update to latest version within the minor version series
+        if key not in latest_versions or version > latest_versions[key][0]:
+            latest_versions[key] = (version, requires_python)
+
+            # Extract the versions and requires_python from the latest_versions dict
+    return dict(latest_versions.values())
+
+
+def main() -> None:
+    """
+    Main function to generate and output the test matrix.
+
+    This function fetches the minimum required Python version for the package
+    ('PACKAGE_NAME'), gets the list of available Python versions starting from
+    that minimum version, and fetches the available versions of the dependency
+    ('DEPENDENCY_NAME') starting from the minimum version required by the package.
+
+    It then computes a matrix of combinations of Python versions and dependency
+    versions that are compatible according to the 'requires_python' specifiers.
+
+    Outputs:
+        A JSON object representing the test matrix, printed to stdout.
+
+    """
+    # Get the minimum required Python version for the package
+    min_python_version = get_package_minimum_python_version(PACKAGE_NAME)
+
+    # Get the available Python versions starting from the minimum version
+    python_versions = get_available_python_versions(min_python_version)
+
+    # Get the minimum required version of the dependency for the package
+    min_dependency_version = get_dependency_minimum_version(PACKAGE_NAME, DEPENDENCY_NAME)
+
+    # Get the available versions of the dependency starting from the minimum version
+    package_versions = get_available_package_versions(
+        DEPENDENCY_NAME,
+        min_dependency_version,
+    )
+
+    include: list[dict[str, str]] = []
+
+    for package_version, requires_python in package_versions.items():
+        specifier_set = SpecifierSet(requires_python)
+
+        for python_version in python_versions:
+            if python_version in specifier_set:
+                # Skip incompatible combinations
+                if python_version >= Version("3.12") and package_version < MINIMUM_DEPENDENCY_VERSION_FOR_PYTHON_3_12:
+                    continue
+
+                include.append(
+                    {
+                        "python": str(python_version),
+                        DEPENDENCY_NAME: str(package_version),
+                    }
+                )
+
+    matrix = {"include": include}
+
+    json.dump(matrix, indent=4, fp=sys.stdout)
```

Comment:
the assignment to `matrix` feels a bit redundant here I guess 🤷🏻

Reply:
I agree, it was really convenient for debugging purposes (i.e. `rich.print(matrix)`)